cpp/src/parquet/arrow/reader.h

```diff
@@ -150,7 +152,17 @@ class PARQUET_EXPORT FileReader {
   /// \returns error Status if row_group_indices contains invalid index
   virtual ::arrow::Status GetRecordBatchReader(
       const std::vector<int>& row_group_indices,
-      std::shared_ptr<::arrow::RecordBatchReader>* out) = 0;
+      std::unique_ptr<::arrow::RecordBatchReader>* out) = 0;
```

Comment:
Do you need to change this? It seems like it would be easier to rather use shared_ptr inside the dataset implementation

Reply:
IMHO this was the less invasive change; we *can* replace all the unique_ptrs in dataset/ with shared_ptrs but it seemed reasonable to just refine the factory here.

Furthermore, unique_ptr seems like the correct output type for any factory function which isn't pulling things out of a cache; they can be implicitly converted to shared_ptr with no cost but the reverse mapping is fraught.